src/charm.py

```diff
@@ -259,14 +274,109 @@ def _on_admin_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app's admin ingress URL: %s", event.url)
 
+        self._update_endpoint_info()
+
     def _on_public_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app's public ingress URL: %s", event.url)
 
+        self._update_endpoint_info()
+
     def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app no longer has ingress")
 
+        self._update_endpoint_info()
+
+    def _on_client_create(self, event: ClientCreateEvent) -> None:
+        if not self._container.can_connect():
+            event.defer()
+            return
+
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            event.defer()
+            return
+
+        client_config = event.to_client_config()
+        client = self._create_client(
+            client_config, metadata=f'{{"relation_id": {event.relation_id}}}'
+        )
+
+        self.oauth.set_client_credentials_in_relation_data(
+            event.relation_id, client["client_id"], client["client_secret"]
+        )
+
+    def _on_client_config_changed(self, event: ClientConfigChangedEvent) -> None:
+        ...
+        # client_config = event.to_client_config()
+        # self._create_client(client_config, metadata=f"{{\"relation_id\": {event.relation_id}}}")
+
+    def _create_client(
```

Comment:
We may want a different class/file for the `hydra` specific calls and have this content in the init. If there are other calls that require using the hydra client they can all be there.